### ISSUE TITLE
Code too large

### ISSUE THREAD

**Issue by manticore-projects** [javacc21/javacc21#47]
After a long back-and-forth I managed to modify the Grammer so JavaCC21 builds a valid Parser in Java.
Unfortunately it won't compile:
```
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project jsqlparser: Compilation failure
/home/are/Documents/src/JSqlParser21/target/generated-sources/javacc/net/sf/jsqlparser/parser/CCJSqlParser.java:[115,51] code too large
```

I am aware of https://javacc.com/2021/03/30/code-too-large-problem-fixed/ but even after reading it 5 times I have not been able to understand, what I am supposed to do for solving this problem.

And why is the Parser generated with JavaCC21 three times the size of the one generated with Legacy JavaCC?

**Comment by manticore-projects** [comment]
I figured, that this "code too large" problem stems from the TokenType EnumSets:
```
first_set$...
follow_set$...
```

So I manually created 2 classes `EnumsFirstSet` and `EnumsFollowSet` and moved all these static EnumSet declarations there.
Each of the class has less than 1000 lines of source code.

But that is not good enough:
```
net/sf/jsqlparser/parser/EnumsFirstSet.java:[329,58] code too large
```

**Comment by manticore-projects** [comment]
A maybe dumb question:
Why can't we just put these EnumSets into a Map instead of defining them as Constants? Maybe something like;

```java
HashMap<String, EnumSet<CCJSqlConstants.TokenType>> enumSetMap = new HashMap<>();
enumSetMap.put("first_set$JSqlParserCC_jjt$5240$5", EnumSet.of(K_MINVALUE, K_NOMINVALUE));
```
instead of
```java
static private final EnumSet<CCJSqlConstants.TokenType> first_set$JSqlParserCC_jjt$5240$5= EnumSet.of(K_MINVALUE, K_NOMINVALUE);
```

**Comment by revusky** [comment]
> After a long back-and-forth I managed to modify the Grammer so JavaCC21 builds a valid Parser in Java.
> Unfortunately it won't compile:
> 
> ```
> Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project jsqlparser: Compilation failure
> /home/are/Documents/src/JSqlParser21/target/generated-sources/javacc/net/sf/jsqlparser/parser/CCJSqlParser.java:[115,51] code too large

Drat. Well, I'll have this fixed fairly quickly.

To be clear on something, the "Code too large" problem mentioned [here](https://javacc.com/2021/03/30/code-too-large-problem-fixed/) that I said was a "thing of the past" refers to something that would occur with some frequency in legacy JavaCC when generating the *Lexer* (a.k.a. TokenManager code). And that genuinely is fixed. But this problem, that in a very large grammar, you can hit the "Code too large" problem generating the *Parser* code... 

The basis of the problem, you see, is that there is a limit to the size of a method in bytecodes, 64K bytecodes in Java, and all static initialization code gets compiled into a single anonymous method. So that single method that is created can easily exceed the aforementioned 64K limit. Well, in *generated* code, it can fairly easily. Just about nobody ever writes code by hand that hits this limit, which is probably why the Java people have never done anything about this. AFAICS, it would be fairly easy for them to quietly fix this by breaking the static initializations that are two large into more than one method... But they never fixed it on their end, so it's a sort of major gotcha that one needs to know about if one is in the business of generating Java code!

So, in this instance, the problem is that, when you define all these EnumSets and initialize them statically, you can hit the Code too large problem generating the parser code, and that is something that is a new bug in JavaCC21 actually, never existed in legacy JavaCC....

```
> 
> I am aware of https://javacc.com/2021/03/30/code-too-large-problem-fixed/ but even after reading it 5 times I have not been able to understand, what I am supposed to do for solving this problem.

Well, *you* are not supposed to have to do anything to solve this problem. As for that blog post, again, it actually refers to a different problem, which is a longstanding problem when generating the code for very big *lexical* specifications and that is fixed in JavaCC21. Unfortunately, it seems that there is now a separate "Code too large" problem for large *syntactic* grammars, which I'll fix shortly.

Of course, the solution is roughly the same. One needs to break the initialization code into multiple pieces. If you generate it in the most natural way, like:

     EnumSet enumSet1 = EnumSet.of(.....);
     EnumSet enumSet2 = EnumSet.of(.....);
     ....
     EnumSet enumSet1000 = EnumSet.of(....);

There, you can hit "code too large" because javac is compiling all that static initialization into a single method. Probably, it's happening in particular because you have a very large number of token types (I'm guessing) so each line is quite a bit of byte code and of course, with a large number of such lines, you hit the 64K method size limit.

> 
> And why is the Parser generated with JavaCC21 three times the size of the one generated with Legacy JavaCC?

Well, I think the main reason is that legacy JavaCC had (and still has) a rather tricky space optimization that I later decided was not worth the candle. Actually, that space optimization was the basis of the bug I talk about [here](https://javacc.com/2020/10/28/a-bugs-life/). (That, by the way, is another longstanding bug in the legacy tool, requiring a single keystroke fix, that I am pretty sure will NEVER be fixed!) After that episode, in my own codebase, I just got rid of the space optimization in question and just generated all the code for a lookahead in the full knowledge that some of it was effectively dead code. I could put that optimization back, it would not be very hard, but... Well, I think that could well account for a lot of the difference in code size. Some other things as well, like if you compare what a grammatical production does by default in JavaCC21 with legacy code, like maintaining the call stack and things like this... But I think the biggest reason might be my removal of the aforementioned space optimization. Truth told, I've been aware that JavaCC21 now generates much bigger code, but I never really gave much priority (so far) to even quantifying the various reasons. Actually, as far as I can tell, code size, on current-day hardware is hardly ever relevant. This 64K bytecode limit, however, is the exception to that! That kind of thing is a real problem! But not to worry, it will be fixed very soon.

**Comment by revusky** [comment]
> I figured, that this "code too large" problem stems from the TokenType EnumSets:
> 
> ```
> first_set$...
> follow_set$...
> ```
> 
> So I manually created 2 classes `EnumsFirstSet` and `EnumsFollowSet` and moved all these static EnumSet declarations there.
> Each of the class has less than 1000 lines of source code.
> 
> But that is not good enough:
> 
> ```
> net/sf/jsqlparser/parser/EnumsFirstSet.java:[329,58] code too large
> ```

Well, the reason that doesn't work is because it doesnt address the problem! The problem is that it creates one big honking method for the static initializations and that big honking method passes the built-in 64K bytecode limit for a single method. So simply breaking off all those initializations into a separate class doesn't solve the problem. You need the initializations to be broken into multiple methods, so that no single method is more than 64K in bytecode.

I don't say the above as a reproach or anything, mind you. I'm just trying to help you to see the exact nature of this problem.

Or, in other problem is not specifically that the XXXParser.java file has too much code. It has too much static initialization code that is compiled into a single method. So when you put all that static initialization code into its own class, you still have the problem!

Well, I put in a fix that should resolve this, but I don't actually have any test code sitting around to check this, so I would appreciate if you got the latest version and see if you still have the problem.

Basically, the very bloody-minded solution that I put in place was that, instead of generating:

     static private final EnumSet<TokenType> enumSetXXX -= EnumSet.of(FOO, BAR, BAZ,....);

I generate:

    static private final EnumSet<TokenType> enumSetXXX = enumSetXXX_init();

    static private EnumSet<TokenType> enumSetXXX_init() {
         return EnumSet.of(FOO, BAR, BAZ,....);
    }

And, granted, that is a bit ugly in the cases (usually) when it is unnecessary. So I might put in a little quick and dirty calculation and only generate the latter code above some threshold, but now I just generate the above code instead of the previous code and it all just seems to work. And, as far as I can see, it won't hit the Code too large problem now.

**Comment by manticore-projects** [comment]
Thank you a lot for the explanations. I am actually aware of many of these details since I wrote a "General Ledger Account Mapping Engine" myself years ago. Up today it is still the "secret sauce" of my company's main product IFRS-VBox. What does it do: it maps any Financial Posting to the the correct Debit and Credit Account in the General Ledger, based on the Instrument's dynamic attributes and properties. Sounds simple, but imagine a Bank with 16 Mill. accounts, producing 10 postings for each every night -- across 3 different Accounting Scopes!

In the end, this comes much closer to a "Parser generator" than I anticipated at that time. It automatically built a similar  Case/If/Then/Else map (dynamically in Groovy, for the better Regex support) and it also ran into "code too large" problems and I had to split it up using a TreeMap structure.

Sometimes it just amazing how the circle closes eventually.
Apologies for so much story telling.

Your commit d15f3eccdce31bd049655abaa41c84d80edaf4df has solved the problem effectively, and my generated JSQLParser compiled. Thank you a lot and cheers!
